### ISSUE TITLE
Current Wallet API Doesn't Allow Merchants to Differentiate Payers

### ISSUE THREAD

**Issue by overlooper** [mimblewimble/grin#2198]
If a merchant were to programmatically implement an application that takes Grin deposits, as of right now there is no variable that can be parsed in the [wallet owner API](https://github.com/mimblewimble/grin/blob/master/doc/api/wallet_owner_api.md#get-retrieve-txs) that would allow the merchant to differentiate a deposit of one payer from another payer's if they deposit the same amount at roughly the same time.

Monero had initially solved this problem with payment IDs that a merchant would provide to the payer before a transaction for them to specify during the transaction in a separate field. This turned out to be a terrible solution that not only compromised privacy at the network level but also created a UX nightmare, where exchanges would ask people to attach a certain payment ID to their transaction, the payers then not doing it due to inexperience or negligence, and the merchants having to sift through the transactions manually to find the payment, resulting in enormous productivity losses all around, users losing their funds due to merchants being unwilling to go through the hassle and a terrible experience for all involved, creating a perception for newcomers that this currency is hard to use and easy to lose.

After Monero had implemented integrated addresses, where the payment ID was baked into the address that the merchants provided to their users for deposits this UX issue was solved at the protocol level, but for years merchants kept using old-style payment IDs, opting to charge their users for finding missing transactions rather than implement the new integrated addresses. Due to backwards compatibility, this problem is [still plaguing Monero even to this day](https://github.com/monero-project/monero/issues/3772).

This is why, in the case of Grin, telling the payer to add a message to the transaction with their merchant portal username is a terrible solution in my opinion.

The ideal solution, in my view, would be a field (let's call it payer-id) that could be specified by the receiver during their stage of the transaction creation, and that the receiver can later parse from a field in the [retrieve_txs](https://github.com/mimblewimble/grin/blob/master/doc/api/wallet_owner_api.md#get-retrieve-txs) method of the owner wallet API when the transaction gets added to the blockchain. This way a merchant would be able to link together all the deposits received from a given user due to that user being provided the same payer-id by the merchant's platform, just like when a payer is given the same integrated address to deposit to in Monero, allowing the merchant to easily scan all transactions from a given payer.

I don't know whether this kind of solution is possible for a MW blockchain, so, in case it isn't, I just want to emphasize that the main requirements for this solution are as follows:

1. It has to require ZERO extra actions from the payer, all the differentiation logic should be done by the merchant.
2. It has to either be something the merchant can specify OR something that a merchant can save before or during a transaction creation that is unique to this transaction or payer that will be later
3. something that the merchant can retrieve during the scanning of an incoming transaction from the blockchain. I.e. a field in the [retrieve_txs](https://github.com/mimblewimble/grin/blob/master/doc/api/wallet_owner_api.md#get-retrieve-txs) method of the owner wallet API.

I also believe this should be prioritized to be implemented before the mainnet launch so as to allow merchants to prepare for launch and, more crucially, avoid uniformity transaction issues in the future (e.g. in Monero the transactions that have a payment ID can have a fingerprint different from the ones that don't).

**Comment by overlooper** [comment]
@yeastplume the key is whether the `receiver_message` field will be available for parsing by the merchant during the owner wallet API `retrieve_txs` call when getting a scanning a new transaction from the blockchain.

I would also request this field be added to the API as soon as possible if this is an easy fix, because the current API is unusable for a business that wants to prepare to accept Grin before the mainnet launch. After all, Grin is positioning itself to be a spending currency, not a store of value, so if businesses can't implement programmatic Grin deposits then the activity will be limited to manual transactions.

**Comment by yeastplume** [comment]
I'm not sure what you're requesting, as no information other that output commits is available when scanning the chain, and you don't learn about incoming transactions via the chain. `retrieve_txs` doesn't pull anything from the chain, it just accesses the transaction log stored locally in the wallet. Since a MW block has no identifying information whatsoever, wallets need to keep a transaction log to link outputs to particular transactions, and you can only store identifying information locally.  

The only way participants can identify themselves during a transaction is via adding information during the transaction building phase, which is what the message field provides. It's up to each party to decide what they want to do with this information and how/whether to store it, but it's available in the slate and transaction object that all parties see before the transaction is sent to the chain.

**Comment by overlooper** [comment]
@yeastplume the merchant can't add the Grin balance to the user's account until they see the tx on the chain, right? So once the merchant sees a tx on the chain, they need to be able to parse some sort of field that will allow them to compare it to the piece of information that links a tx to a user to credit their account.

**Comment by sesam** [comment]
overlooper just like with most currencies, the merchant wallet is the first to see the transaction, and the process for merchants is that the finalized transaction is sent to the chain _from the merchants wallet_. Thus that wallet does not need to wait for the tx to be mined. But before finalizing the deposit (making it available / shipping goods) the merchant still needs to wait for ~10 confirmations on the chain.

**Comment by overlooper** [comment]
@yeastplume it does sound it would cover it, apart from the specifying the amount part. Ideally, the amount should be specified by the sender. Also, is this workflow currently implementable with the available wallet API?

**Comment by yeastplume** [comment]
Right, so doing this is on the radar, but no, it can't be done in the current API. The functionality is easily available with the core wallet libraries that are already there, but it still needs to be put together, tests written, workflows added to command line wallet, tested some more, etc. Given the focus right now is ensuring everything is robust for mainnet launch, I don't think we can promise to get it in before then.

However, I'd be comfortable saying it should be coming soon after, since it makes a lot of sense to get this workflow option in place as soon as possible. Will definitely go into the feature list for the point release after mainnet.

**Comment by garyyu** [comment]
I link the related discussion in Gitter lobby happened at 21th Dec. 2018:  https://gitter.im/grin_community/Lobby?at=5c1ce1dfd945b96882d390f1 .

And a related topic in forum about `recipient initiates / invoice` transaction requirement:  https://www.grin-forum.org/t/reverse-transaction-building/482

Since we still have a few bugs to be fixed to make Grin Sever/Wallet stable/secure enough before the mainnet launching, I agree it's not practical to get this nice `recipient initiates / invoice` feature implemented before mainnet.

Now I'm thinking another way to implement this `merchant payment id` feature, which is quite simple and no need a big work, just a bit enhancement on current `sender initiates` transaction implementation, and compatible to current `sender initiates` transaction.

A draft description of this idea:
```
Alice                                        Merchant
 |
 | slate  ---------------------------------> (check if a conformed format 'merchant_payment_id' is present in the user 'message' field)
 |
 | 'merchant_payment_id' <-----------------  if no, generate an unique 'merchant_payment_id' and send back to Alice, and save to local database of merchant
 |  
 | slate' ---------------------------------> verify the message signature and check & get the conformed format `merchant_payment_id`
 | 
 | (remaining parts are same as current implementation, need map the `merchant_payment_id` to the transaction `output` in local database of merchant, so as to check the payment status on chain)
 |
```
Above `slate'`  is still the `slate` in 1st step but with 'merchant_payment_id' in message and with a signature.

And if a merchant don't want to use/check this `merchant_payment_id`, everything will be same as now.

**Comment by seibelj** [comment]
A clever way to solve this would be to allow a special HTTP header, like `X-Payer-Id`.

This would be set in the wallet configuration and optionally in a flag for the send.

There should also be a wallet configuration to reject any sends that hit the foreign wallet ID that don't have this set.

The exchange provides each user a unique Payer-ID, the customer sets it or provides it during a send, and when the user calls the exchange's foreign wallet endpoint it gets passed along (or rejected if the customer forgot to set it).

Then any deposits that hit the exchanges main endpoint (http://grin.exchange.com/) have their X-Payer-ID header parsed and it's matched to the user.

Because all communication is outside of the blockchain, and we have HTTP endpoints, we can do all sorts of simple tricks to improve flows. This is one suggestion.

**Comment by BitHighlander** [comment]
perhaps an invoiceId.exchange.com:13420 with a reverse proxy would be a way to attach a invoice without any involvement/actions from the user.

When a exchange hands a destination to a customer that invoice can be generated. 

imho any protocol that allows a customer to mess up a transaction will turn into a nightmare of support. 

This reverse proxy project could be devolved in a optional stand alone way that sits in front of the wallet.

**Comment by lehnberg** [comment]
The reverse transaction building flow discussed above is currently implemented (in a rudimentary form) in wallet713 as the `invoice` command:
https://github.com/vault713/wallet713/blob/master/docs/usage.md#using-invoice

@overlooper why do you want the amount specified by the sender? Wouldn't it be preferred that the recipient specifies the amount as part of issuing the invoice, and the sender just confirms that indeed, they are okay to pay it? Are you thinking of exchange deposit flows? Can you help me understand how you would want it to work in your ideal scenario, can you describe how the flow would look?
